cohortextractor/pandas_utils.py

```diff
@@ -0,0 +1,111 @@
+from pathlib import Path
+
+import pandas
+
+
+def dataframe_to_file(df, filename):
+    extension = Path(filename).suffix.lstrip(".")
+    if extension == "csv":
+        df.to_csv(filename, index=False)
+    elif extension == "feather":
+        df.to_feather(filename)
+    elif extension == "dta":
+        df.to_stata(filename, write_index=False)
+    else:
+        raise RuntimeError(f"Unsupported extension: {extension}")
+
+
+def dataframe_from_rows(covariate_definitions, rows):
+    """
+    Create a DataFrame from an iterator of rows
+    """
+    convertors = {
+        name: get_pandas_convertor(funcname, **kwargs)
+        for name, (funcname, kwargs) in covariate_definitions.items()
+    }
+    # `patient_id` isn't included in `covariate_definitions` so we need to add
+    # it manually
+    convertors["patient_id"] = identity
+
+    # First row is expected to be header row
+    headers = next(rows)
+    convertor_funcs = [convertors[column] for column in headers]
+
+    def convert_row(row):
+        return [convertor(value) for (convertor, value) in zip(convertor_funcs, row)]
+
+    data = map(convert_row, rows)
+    df = pandas.DataFrame(data, columns=headers)
+    # Change categorical columns from arrays of ints to proper Pandas
+    # Categorical objects
```

Comment:
Fortunately this isn't actually a second pass. Categoricals are just implemented as arrays of ints together with a reference array giving the categories. We build the int array as we go along and then use `Categorical.from_codes` to turn this directly into a Categorical array without needing to re-process it. It's certainly more faff, but shouldn't be a major performance issue.

Reply:
Yes, I meant from a faff PoV rather than performance.  I really wanted to make the comment because I realised that the categories really should be part of the contract *anyway*. I think.